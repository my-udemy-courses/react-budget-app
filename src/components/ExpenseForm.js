import React from 'react'
import moment from 'moment'
import { SingleDatePicker } from 'react-dates'

export default class ExpenseForm extends React.Component {
    constructor(props) {
        super(props)
        const { expense = null } = props
        this.state = {
            description: expense ? expense.description : '',
            note: expense ? expense.note : '',
            amount: expense ? (expense.amount / 100).toString() : '',
            createdAt: expense ? moment(expense.createdAt) : moment(),
            calendarFocused: false,
            error: ''
        }
    }
    onDescriptionChange = (e) => {
        const description = e.target.value
        this.setState(() => ({ description }))
    }
    onNoteChange = (e) => {
        const note = e.target.value
        this.setState(() => ({ note }))
    }
    onAmountChange = (e) => {
        const amount = e.target.value

        // Only allow digits and with two decimal places at most
        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => ({ amount }))
        }
    }
    onDateChange = (createdAt) => {
        if (createdAt) {
            this.setState(() => ({ createdAt }))
        }

    }
    onCalendarFocusChange = ({ focused }) => {
        this.setState(() => ({ calendarFocused: focused }))
    }
    onSubmit = (e) => {
        e.preventDefault()

        if (!this.state.description || !this.state.amount) {
            this.setState(() => ({ error: 'Bitte mindestens Beschreibung und Höhe der Ausgabe angeben.' }))
        } else {
            this.setState(() => ({ error: '' }))
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount) * 100,
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note
            })
        }
    }
    render() {
        return (
            <form className="form" onSubmit={this.onSubmit}>
                { this.state.error && <p className="form__error">{this.state.error}</p> }
                <input
                    type="text"
                    placeholder="Beschreibung"
                    autoFocus
                    className="text-input"
                    value={this.state.description}
                    onChange={this.onDescriptionChange}
                />
                <input
                    type="text"
                    placeholder="Höhe der Ausgabe in €"
                    className="text-input"
                    value={this.state.amount}
                    onChange={this.onAmountChange}
                />
                <SingleDatePicker
                    date={this.state.createdAt}
                    onDateChange={this.onDateChange}
                    focused={this.state.calendarFocused}
                    onFocusChange={this.onCalendarFocusChange}
                    numberOfMonths={1}
                    isOutsideRange={() => false} // every day is allowed aka in range
                />
                <textarea
                    placeholder="Notiz (optional)"
                    className="textarea"
                    value={this.state.note}
                    onChange={this.onNoteChange}
                >
                </textarea>
                <div>
                    <button className="button">{ this.props.buttonLabel ? this.props.buttonLabel : 'Speichern' }</button>
                    { this.props.deleteButton && this.props.deleteButton }
                    { this.props.cancelButton && this.props.cancelButton }
                </div>
            </form>
        )
    }
}
