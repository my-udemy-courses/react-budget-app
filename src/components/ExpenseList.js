import React from 'react'
import { connect } from 'react-redux'
import ExpenseListItem from './ExpenseListItem'
import selectExpenses from '../selectors/expenses'

export const ExpenseList = ({expenses}) => {
    return (
    <div className="content-container">
        <div className="list-header">
            <div className="show-for-mobile">Ausgaben</div>
            <div className="show-for-desktop">Ausgabe</div>
            <div className="show-for-desktop">HÃ¶he</div>
        </div>
        <div className="list-body">
            {
                expenses.length === 0 ? (
                    <div className="list-item list-item--message">
                        <span>Keine Ausgaben</span>
                    </div>
                ) : (
                    expenses.map(expense => (
                        <ExpenseListItem key={expense.id} {...expense} />
                    ))
                )
            }
        </div>
    </div>)
}

const mapStateToProps = (state) => ({
    expenses: selectExpenses(state.expenses, state.filters)
})

export default connect(mapStateToProps)(ExpenseList)
