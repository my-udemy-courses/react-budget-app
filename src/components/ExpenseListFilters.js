import React from 'react'
import { connect } from 'react-redux'
import { DateRangePicker } from 'react-dates'
import { setTextFilter, sortByAmount, sortByDate, setDescending, setStartDate, setEndDate } from '../actions/filters'

export class ExpenseListFilters extends React.Component {
    state = {
        calendarFocused: null
    }
    onDatesChange = ({ startDate, endDate }) => {
        this.props.setStartDate(startDate)
        this.props.setEndDate(endDate)
    }
    onFocusChange = (calendarFocused) => {
        this.setState(() => ({ calendarFocused }))
    }
    onTextChange = (e) => {
        this.props.setTextFilter(e.target.value)
    }
    onSortChange = (e) => {
        const val = e.target.value
                    
        if (val === 'date') {
            this.props.sortByDate()
        } else {
            this.props.sortByAmount()
        }
    }
    render() {
        return (
            <div>
                <input 
                    className="text-filter"
                    type="text" 
                    value={this.props.filters.text} 
                    onChange={this.onTextChange}
                />
        
                <select onChange={this.onSortChange}>
                    <option value="date">Date</option>
                    <option value="amount">Amount</option>
                </select>
        
                <label>
                    <input 
                        className="sort-direction"
                        type="checkbox" 
                        id="ascending-checkbox" 
                        onChange={(e) => {
                            this.props.setDescending(!e.target.checked)
                        }} 
                        checked={!this.props.filters.descending}
                    />
                    Ascending
                </label>
        
                <DateRangePicker
                    startDate={this.props.filters.startDate}
                    endDate={this.props.filters.endDate}
                    onDatesChange={this.onDatesChange}
                    focusedInput={this.state.calendarFocused}
                    onFocusChange={this.onFocusChange}
                    numberOfMonths={1}
                    isOutsideRange={(date) => false}
                    showClearDates={true}
                />
        
            </div>
        )        
    }
}

const mapStateToProps = (state) => ({
    filters: state.filters
})

const mapDispatchToProps = (dispatch) => ({
    setTextFilter: (text) => dispatch(setTextFilter(text)),
    sortByDate: () => dispatch(sortByDate()),
    sortByAmount: () => dispatch(sortByAmount()),
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate)),
    setDescending: (isDescending) => dispatch(setDescending(isDescending)),
})

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters)