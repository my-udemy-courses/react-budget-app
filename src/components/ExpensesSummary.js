import React from 'react'
import { connect } from 'react-redux'
import { Link } from 'react-router-dom'
import numeral from 'numeral'
import selectExpenses from '../selectors/expenses'
import selectExpensesTotal from '../selectors/expenses-total'

export const ExpensesSummary = (props) => {
    const expenseWord = props.expensesCount === 1 ? 'Ausgabe' : 'Ausgaben'
    const formattedTotal = numeral(props.expensesTotal / 100).format('0,0.00') + ' €'

    return (
        <div className="page-header">
            <div className="content-container">
                <h1 className="page-header__title"><span>{props.expensesCount}</span> {expenseWord}, insgesamt <span>{formattedTotal}</span></h1>
                <div className="page-header__actions">
                    <Link className="button" to="/create">Ausgabe hinzufügen</Link>
                </div>
            </div>
        </div>
    )
}

const mapStateToProps = (state) => {
    const expenses = selectExpenses(state.expenses, state.filters)
    return {
        expensesCount: expenses.length,
        expensesTotal: selectExpensesTotal(expenses)
    }
}

export default connect(mapStateToProps)(ExpensesSummary)
