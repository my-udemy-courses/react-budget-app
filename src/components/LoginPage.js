import React from 'react'
import { connect } from 'react-redux'
import { startLogin } from '../actions/auth'

export const LoginPage = ({ startLogin }) => (
    <div className="box-layout">
        <div className="box-layout__box">
            <h1 className="box-layout__title">Ausgaben Checker</h1>
            <p>Hilft dabei, Ausgaben im Blick zu behalten.</p>
            <button className="button" onClick={startLogin}>Login mit Google</button>
        </div>
    </div>
)

const mapDispatchToProps = (dispatch) => ({
    startLogin: () => dispatch(startLogin())
})

export default connect(undefined, mapDispatchToProps)(LoginPage)
