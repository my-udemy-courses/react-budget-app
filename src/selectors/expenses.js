import moment from 'moment'

// Get visible expenses => Filter Logic
export default (expenses, { text, sortBy, startDate, endDate, descending }) => {
    return expenses.filter(({ description, createdAt }) => {
        const createdAtMoment = moment(createdAt)
        const startDateMatch = startDate ? createdAtMoment.isSameOrAfter(startDate, 'day') : true
        const endDateMatch = endDate ? createdAtMoment.isSameOrBefore(endDate, 'day'): true
        const textMatch = description.toLowerCase().includes(text.toLowerCase())

        return startDateMatch && endDateMatch && textMatch
    }).sort((a, b) => {
        if (sortBy === 'date') {
            if (descending){
                return a.createdAt < b.createdAt ? 1 : -1
            } else {
                return a.createdAt < b.createdAt ? -1 : 1
            }
        } 
        else if (sortBy === 'amount') {
            if (descending) {
                return a.amount < b.amount ? 1 : -1
            } else {
                return a.amount < b.amount ? -1 : 1
            }
        }
    })
}
