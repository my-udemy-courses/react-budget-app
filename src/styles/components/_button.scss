.button {
    background: $blue;
    border: none;
    color: white;
    display: inline-block;
    font-size: $font-size-large;
    font-weight: 300;
    line-height: 1;
    padding: $s-size;
    text-decoration: none;
}

.button--link {
    background: none;
}

.button--dark-link {
    @extend .button--link;
    color: $blue;
}

.button--secondary {
    background: #888888;
}
