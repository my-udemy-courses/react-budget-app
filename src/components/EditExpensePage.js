import React from 'react'
import { connect } from 'react-redux'
import ExpenseForm from './ExpenseForm'
import { editExpense, startRemoveExpense } from '../actions/expenses'

export class EditExpensePage extends React.Component {
    onSubmit = (updatedExpense) => {
        this.props.editExpense(this.props.expense.id, updatedExpense)
        this.props.history.push('/')
    }
    onRemove = () => {
        this.props.startRemoveExpense(this.props.expense.id)
        this.props.history.push('/')    
    }
    render() {
        return (
            <div>
                <ExpenseForm 
                    expense={this.props.expense}
                    onSubmit={this.onSubmit}
                    buttonLabel={'Update Expense'}
                />
                <button onClick={this.onRemove}>Remove Expense</button>
            </div>
        )
    }
}

const mapStateToProps = (state, props) => ({
    expense: state.expenses.find((expense) => expense.id === props.match.params.id)
})

const mapDispatchToProps = (dispatch) => ({
    editExpense: (id, updatedExpense) => dispatch(editExpense(id, updatedExpense)),
    startRemoveExpense: (id) => dispatch(startRemoveExpense(id)),
})

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage)