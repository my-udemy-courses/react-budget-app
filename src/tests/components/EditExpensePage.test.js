import React from 'react'
import { shallow } from 'enzyme'
import { EditExpensePage } from '../../components/EditExpensePage'
import expenses from '../fixtures/expenses'

let startRemoveExpenseSpy, startEditExpenseSpy, historySpy, wrapper

beforeEach(() => {
    startEditExpenseSpy = jest.fn()
    startRemoveExpenseSpy = jest.fn()
    historySpy = { push: jest.fn() }

    wrapper = shallow(<EditExpensePage
        expense={expenses[0]} 
        startEditExpense={startEditExpenseSpy} 
        startRemoveExpense={startRemoveExpenseSpy}
        history={historySpy}
    />)
})

test('should render EditExpensePage', () => {
    expect(wrapper).toMatchSnapshot()
})

test('should handle editExpense', () => {
    wrapper.find('ExpenseForm').prop('onSubmit')(expenses[1])
    expect(historySpy.push).toHaveBeenLastCalledWith('/')
    expect(startEditExpenseSpy).toHaveBeenLastCalledWith(expenses[0].id, expenses[1])
})

test('should handle startRemoveExpenseSpy', () => {
    wrapper.find('button').simulate('click')
    expect(historySpy.push).toHaveBeenLastCalledWith('/')
    expect(startRemoveExpenseSpy).toHaveBeenLastCalledWith(expenses[0].id)
})
