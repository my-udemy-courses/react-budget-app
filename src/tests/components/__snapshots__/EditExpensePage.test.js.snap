// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EditExpensePage 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <h1
        className="page-header__title"
      >
        Ausgabe bearbeiten
      </h1>
    </div>
  </div>
  <div
    className="content-container"
  >
    <ExpenseForm
      buttonLabel="Speichern"
      cancelButton={
        <button
          className="button button--dark-link"
          onClick={[Function]}
        >
          Abbrechen
        </button>
      }
      deleteButton={
        <button
          className="button button--secondary"
          onClick={[Function]}
        >
          LÃ¶schen
        </button>
      }
      expense={
        Object {
          "amount": 195,
          "createdAt": 0,
          "description": "Gum",
          "id": "1",
          "note": "",
        }
      }
      onSubmit={[Function]}
    />
  </div>
</div>
`;
